test/scale/scale_tester.go

```diff
@@ -0,0 +1,358 @@
+	nodeutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/node"
+	podutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/pod"
+	volutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/volume"
+)
+
+type PodSchedulingWithPVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverMultiplePVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+func (t *PodSchedulingWithPVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	totalNumberOfPods := t.Replicas
+	var oneMinPodsRunning,
+		twoMinPodsRunning,
+		threeMinPodsRunning,
+		fiveMinPodsRunning,
+		tenMinPodsRunning = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	var oneMinPodsDeleted,
+		twoMinPodsDeleted,
+		threeMinPodsDeleted,
+		fiveMinPodsDeleted,
+		tenMinPodsDeleted = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	startScaleOut := time.Now()
+	// Get the list of available nodes for scheduling the pod
+	nodes := nodeutil.ListNodeNames(client)
+	if len(nodes) < 3 {
+		ginkgo.Skip(fmt.Sprintf("need at least 3 nodes to verify the test case. Current node count is %d", len(nodes)))
+	}
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	statefulSet, cleanupStatefulSet := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+	for i := range cleanupStatefulSet {
+		i := i
+		defer cleanupStatefulSet[i]()
+	}
+
+	statefulSet.CreateWithoutWaiting()
+	numberOfRunningPods := 0
+	for {
+		numberOfRunningPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"app": "azuredisk-volume-tester"}, string(v1.PodRunning))
+		if numberOfRunningPods >= totalNumberOfPods || time.Since(startScaleOut) > 30*time.Minute {
+			break
+		}
+		timePassed := time.Since(startScaleOut)
+		switch {
+		case timePassed <= 1*time.Minute:
+			oneMinPodsRunning = numberOfRunningPods
+		case timePassed <= 2*time.Minute:
+			twoMinPodsRunning = numberOfRunningPods
+		case timePassed <= 3*time.Minute:
+			threeMinPodsRunning = numberOfRunningPods
+		case timePassed <= 5*time.Minute:
+			fiveMinPodsRunning = numberOfRunningPods
+		case timePassed <= 10*time.Minute:
+			tenMinPodsRunning = numberOfRunningPods
+		}
+		time.Sleep(30 * time.Second)
```

Comment:
Using `Sleep(30 * time.Seconds)` in a loop make it likely that we're grossly under-reporting the ready pods for each time interval. The time at each iteration will "slip" by the time take to execute `CountAllPodsMatchingLabel` meaning that the first bucket will only contain the pods ready within the first 30 seconds.

A more reliable approach would use a ticker to iterate at a specific interval. See [`NewTicker`](https://pkg.go.dev/time#NewTicker) for an example.

Reply:
The main reason for using `Sleep` in the loop was to avoid issuing too many List requests to the apiserver and getting throttled which can affect the performance of the whole test. But I agree, that the drawback here is that we  underreport the number of the pods in each bucket.

Reply:
You can achieve the same rate limiting with `Ticker` but without the time drift from `Sleep`.

Reply:
Oh, I see what you mean. I'll update it tomorrow!

Reply:
Modified the code to use Ticker